test/regtest.py

```diff
@@ -69,78 +79,171 @@ def metadata(file):
+    if passed:
+      return PASSED
+    elif timedout:
+      return TIMEDOUT
+    elif unknown:
+      return UNKNOWN
+    else:
+      return FAILED
+
+test_results = []
+def tally_result(result):
+    """
+    Called whenever a worker has finished its process. Only the main process
+    will modify test_results.
+    """
+    test_results.append(result)
+
+def main():
+    """
+    Main entry point for the test suite.
+    """
+    t0 = time.time()
+    num_cpus = multiprocessing.cpu_count()
+
+    # configure the CLI
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--exhaustive", help="be exhaustive", action="store_true")
+    parser.add_argument("--threads", action="store", dest="n_threads", default=num_cpus, type=int,
+                        help="execute regressions using the selected number of threads in parallel")
+    parser.add_argument("--log", action="store", dest="log_level", default="DEBUG", type=str,
+                        help="sets the logging level (DEBUG, INFO, WARNING)")
+    parser.add_argument("--output_log", action="store", dest="log_path", type=str,
+                        help="sets the output log path. (std out by default)")
+    args = parser.parse_args()
+
+
+    # configure the logging
+    log_format = ''
+    log_level = logging.DEBUG
+
+    # add more log levels later (if needed)
+    if args.log_level.upper() == "INFO":
+      log_level = logging.INFO
+    elif args.log_level.upper() == "WARNING":
+      log_level = logging.WARNING
+
+    # if the user supplied a log path, write the logs to that file.
+    # otherwise, write the logs to std out.
+    if args.log_path:
+      logging.basicConfig(filename=args.log_path, format=log_format, level=log_level)
+    else:
+      logging.basicConfig(format=log_format, level=log_level)
+
+    logging.debug("Creating Pool with '%d' Workers" % args.n_threads)
+    p = ThreadPool(processes=args.n_threads)
+
+    # start the tests
+    logging.info("Running regression tests...")
+
+    # start processing the tests.
+    results = []
+    for test in glob.glob("./**/*.c"):
+      # get the meta data for this test
+      meta = metadata(test)
+
+      if meta['skip']:
+        continue
+
+      if meta['skip'] != False and not args.exhaustive:
+        continue
+
+      # build up the subprocess command
+      cmd = ['smackverify.py', test]
+      cmd += ['--time-limit', str(meta['time-limit'])]
+      cmd += ['-o', test + '.bpl']
+      cmd += meta['flags']
+
+      for memory in meta['memory'][:100 if args.exhaustive else 1]:
+        cmd += ['--mem-mod=' + memory]
+
+        for verifier in meta['verifiers'][:100 if args.exhaustive else 1]:
+          cmd += ['--verifier=' + verifier]
+          r = p.apply_async(process_test, args=(cmd, test, memory, verifier, meta,), callback=tally_result)
+          results.append(r)
+
+    try:
+      # keep the main thread active while there are active workers
+      for r in results:
```

Comment:
Is this loop superior to the following?
````
while True:
    time.sleep(1)
````
in particular, will a keyboard interrupt work if the first result is not terminating any time soon?